src/cython/sktda/metrics.py

```diff
@@ -0,0 +1,226 @@
+"""
+@author: Mathieu Carriere
+All rights reserved
+"""
+
+import sys
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.metrics import pairwise_distances
+try:
+    from gudhi import bottleneck_distance
+    USE_GUDHI = True
+except ImportError:
+    USE_GUDHI = False
+    print("Gudhi not found: BottleneckDistance will return null matrix, and exact SlicedWassersteinDistance not available")
+
+#############################################
+# Metrics ###################################
+#############################################
+
+class SlicedWassersteinDistance(BaseEstimator, TransformerMixin):
+    """
+    This is a class for computing the sliced Wasserstein distance matrix from a list of persistence diagrams. The Sliced Wasserstein distance is computed by projecting the persistence diagrams onto lines, comparing the projections with the 1-norm, and finally integrating over all possible lines. See http://proceedings.mlr.press/v70/carriere17a.html for more details. 
+    """
+    def __init__(self, num_directions=10):
+        """
+        Constructor for the SlicedWassersteinDistance class.
+
+        Attributes:
+            num_directions (int): number of lines to sample uniformly from [-pi,pi] in order to approximate and speed up the distance computation (default 10). 
+        """
+        self.num_directions = num_directions
+        thetas = np.linspace(-np.pi/2, np.pi/2, num=self.num_directions+1)[np.newaxis,:-1]
+        self.lines_ = np.concatenate([np.cos(thetas), np.sin(thetas)], axis=0)
+
+    def fit(self, X, y=None):
+        """
+        Fit the SlicedWassersteinDistance class on a list of persistence diagrams: persistence diagrams are projected onto the different lines. The diagrams themselves and their projections are then stored in a numpy array called **diagrams**.
+
+        Parameters:
+            X (list of n x 2 numpy arrays): input persistence diagrams.
+            y (n x 1 array): persistence diagram labels (unused).
+        """
+        self.diagrams_ = X
+        self.approx_ = [np.matmul(X[i], self.lines_) for i in range(len(X))]
+        diag_proj = (1./2) * np.ones((2,2))
+        self.approx_diag_ = [np.matmul(np.matmul(X[i], diag_proj), self.lines_) for i in range(len(X))]
+        return self
+
+    def transform(self, X):
+        """
+        Compute all sliced Wasserstein distances between the persistence diagrams that were stored after calling the fit() method, and a given list of (possibly different) persistence diagrams.
+
+        Parameters:
+            X (list of n x 2 numpy arrays): input persistence diagrams.
+
+        Returns:
+            Xfit (numpy array of shape (number of diagrams in **diagrams**) x (number of diagrams in X)): matrix of pairwise sliced Wasserstein distances.
```

Comment:
I had a look at scikit-learn, and in particular at functions that return a matrix of pairwise distances/kernels (sklearn.metrics.pairwise.sigmoid_kernel, sklearn.metrics.pairwise.cosine_similarity, etc) and they do not use fit/transform for those, just plain functions. Should we do the same?

Reply:
I agree, usually `sklearn` uses plain functions for this kind of computation. The reason I sticked to a class with `fit` and `transform` methods is because it allows me to treat k-NN with bottleneck distances in the same way than other vector or kernel methods. See the notebook http://localhost:8888/notebooks/3DSeg/3DSegTDA.ipynb for an example

Reply:
That seems to make sense. Which confuses me: why doesn't sklearn use its own fit/transform interface, if that would make its use more convenient?